File: performanceComparisonBioinformatics/python/pairedAlignment.py
import math
import numpy
import time
import sys
cadena1=''
cadena2=''
gap=2
mut=1
score=float(0)
def dinamica(cadena1,cadena2):
	##
	##Construccion matriz dinamica
	##

	w = len(cadena2)+1;
	h=len(cadena1)+1;
	matrizDinamica = numpy.zeros([h,w])
	#print(str(w)+"-----"+str(h))
	##llenarScoresBase

	for i in range (0,len(matrizDinamica)):
		matrizDinamica[i][0]=i*gap
		#print(matrizDinamica[i][0])

	for i in range (0,len(matrizDinamica[0])):
		matrizDinamica[0][i]=i*gap
		#print(matrizDinamica[0][i])
		
	for i in range(1,len(matrizDinamica)):
		for j in range (1,len(matrizDinamica[i])):
			upVal=matrizDinamica[i-1][j]
			leftVal=matrizDinamica[i][j-1]
			diagVal=matrizDinamica[i-1][j-1]
			costUp=upVal+gap
			costLeft=leftVal+gap
			costDiag=diagVal
			if(cadena1[i-1]!=cadena2[j-1]):
				costDiag=costDiag+mut
			matrizDinamica[i][j]=min(min(costDiag,costUp),costLeft)

	score=matrizDinamica[len(matrizDinamica)-1][len(matrizDinamica[0])-1]			
	#print(len(cadena1))
	#print(len(cadena2))
	#print(matrizDinamica)
	print(score)
	#print(matrizDinamica.shape)
	#print(matrizDinamica[0][10])
	#print(matrizDinamica[10][0])
	#print(matrizDinamica)
	#print(len(cadena1))

files=sys.argv
file1=files[1]
file2=files[2]
with open("fastaFiles/"+file1,"r")as f:
	i=0
	for l in f:
		if(i!=0):
			cadena1+=l
		else:
			i+=1
with open("fastaFiles/"+file2,"r")as f2:
	i=0
	for l in f2:
		if(i!=0):
			cadena2+=l
		else:
			i+=1
			
d1=cadena1.split("\n")
d2=cadena2.split("\n")
cad1=""
cad2=""
for d in d1:
	cad1+=d
	
for d in d2:
	cad2+=d
dinamica(cad1,cad2)

File: performanceComparisonBioinformatics/javascript/pairedAlignment.js
var XMLHttpRequest = require("xmlhttprequest").XMLHttpRequest;
var now = require("performance-now");
var dna1='';
var dna2='';
gap=2;
mut=1;
function readTextFile1(file)
{
    var rawFile = new XMLHttpRequest();
    rawFile.open("GET", file, false);
    rawFile.onreadystatechange = function ()
    {
        if(rawFile.readyState === 4)
        {
            if(rawFile.status === 200 || rawFile.status == 0)
            {
                var allText = rawFile.responseText;
				dna1=allText;
				return allText;
            }
        }
    }
    rawFile.send(null);
}

function readTextFile2(file)
{
    var rawFile = new XMLHttpRequest();
    rawFile.open("GET", file, false);
    rawFile.onreadystatechange = function ()
    {
        if(rawFile.readyState === 4)
        {
            if(rawFile.status === 200 || rawFile.status == 0)
            {
                var allText = rawFile.responseText;
				dna2=allText;
				return allText;
            }
        }
    }
    rawFile.send(null);
}

function dinamica (cadena1, cadena2){
	var matrizDinamica=new Array(cadena1.length+1)
	for(var i=0;i<matrizDinamica.length;i++){
		matrizDinamica[i]= new Array(cadena2.length+1)
	}
		
	//Llenar scores de los extremos
	for (var i = 0; i < matrizDinamica.length; i++) {
		matrizDinamica[i][0]=i*gap;
	}
	for (var i = 1; i < matrizDinamica[0].length; i++) {
		matrizDinamica[0][i]=i*gap	;
	}

	for (var i = 1; i < matrizDinamica.length; i++) {
		for (var j = 1; j < matrizDinamica[i].length; j++) {
			var upVal=matrizDinamica[i-1][j];
			var leftVal=matrizDinamica[i][j-1];
			var diagVal=matrizDinamica[i-1][j-1];
			
			var costUp=0;
			var costDiag=diagVal;
			var costLeft=0;
			
			costUp=upVal+gap;
			costLeft=leftVal+gap;
			if(!(cadena1.charAt(i-1)==cadena2.charAt(j-1))) {
				costDiag+=mut;
			}
			matrizDinamica[i][j]=Math.min(Math.min(costDiag, costLeft), costUp);
		}
	}
	score=matrizDinamica[cadena1.length][cadena2.length];
	console.log(score)
}

var myArgs = process.argv.slice(2);
readTextFile1("file:///Users/Jularenas/Desktop/thesis/LanguagePerformanceComparison-Bioinformatics/performanceComparisonBioinformatics/javascript/fastaFiles/"+myArgs[0]);
readTextFile2("file:///Users/Jularenas/Desktop/thesis/LanguagePerformanceComparison-Bioinformatics/performanceComparisonBioinformatics/javascript/fastaFiles/"+myArgs[1]);

var d1=dna1.split("\n");
var d2=dna2.split("\n");
var cad1="";
var cad2="";
for(var i=0;i<d1.length;i++){
	cad1+=d1[i];
}
for(var i=0;i<d2.length;i++){
	cad2+=d2[i];
}
//console.log(cad1);
//console.log(cad2);
dinamica(cad1,cad2);
